modules/extensions/src/terrain/terrain-effect.ts

```diff
@@ -0,0 +1,238 @@
+      this.initialize(gl);
+      for (const layer of opts.layers) {
+        // Force the terrain layer (and its descendents) to rebuild their models with the new shader
+        if (layer.props.operation.includes('terrain')) {
+          layer.setChangeFlags({extensionsChanged: true});
+        }
+      }
+    }
+
+    const {viewports, pass} = opts;
+    const isPicking = pass.startsWith('picking');
+    this.isPicking = isPicking;
+
+    // TODO - support multiple views?
+    const viewport = viewports[0];
+    const layers = (isPicking ? this.terrainPickingPass : this.terrainPass).getRenderableLayers(
+      viewport,
+      opts as TerrainPickingPassRenderOptions
+    );
+
+    const terrainLayers = layers.filter(l => l.props.operation.includes('terrain'));
+    if (terrainLayers.length === 0) {
+      return;
+    }
+
+    if (!isPicking) {
+      const offsetLayers = layers.filter(l => l.state.terrainFittingMode === 'offset');
+      if (offsetLayers.length > 0) {
+        this._renderHeightMap(gl, terrainLayers, viewport, opts);
+      }
+    }
+
+    const drapeLayers = layers.filter(l => l.state.terrainFittingMode === 'drape');
+    this._updateTerrainCovers(terrainLayers, drapeLayers, viewport, opts);
+  }
+
+  getModuleParameters(layer: Layer): TerrainModuleSettings {
+    const terrainCover = this.terrainCovers.get(layer.id);
+    const {terrainFittingMode} = layer.state;
+
+    let terrainCoverTexture: Texture2D | null;
+    if (!terrainCover) {
+      terrainCoverTexture = null;
+    } else if (this.isPicking) {
+      terrainCoverTexture = terrainCover.getPickingFramebuffer();
+    } else {
+      terrainCoverTexture = terrainCover.getRenderFramebuffer();
+    }
+
+    return {
+      heightMap: this.heightMap?.color,
+      dummyHeightMap: this.dummyHeightMap,
+      terrainCover: terrainCoverTexture,
+      terrainCoverBounds: terrainCover?.commonBounds,
+      useTerrainHeightMap: terrainFittingMode === 'offset',
+      terrainSkipRender: terrainFittingMode === 'drape' || !layer.props.operation.includes('draw')
+    };
+  }
+
+  cleanup(): void {
+    if (this.dummyHeightMap) {
+      this.dummyHeightMap.delete();
+      this.dummyHeightMap = undefined;
+    }
+
+    if (this.heightMap) {
+      this.heightMap.color.delete();
+      this.heightMap.delete();
+      this.heightMap = undefined;
+    }
+
+    for (const terrainCover of this.terrainCovers.values()) {
+      terrainCover.delete();
+    }
+    this.terrainCovers.clear();
+  }
+
+  private _renderHeightMap(
+    gl: WebGLRenderingContext,
+    terrainLayers: Layer[],
+    viewport: Viewport,
+    opts: PreRenderOptions
+  ) {
+    if (!this.heightMap) {
+      // Not supported
+      return;
+    }
+
+    this.heightMap.resize(viewport);
+
+    this.terrainPass.renderHeightMap(this.heightMap, {
+      ...opts,
+      pass: 'terrain-height-map',
+      viewports: [viewport],
+      layers: terrainLayers,
```

Comment:
Another worry about updating the viewport is that slight changes in the viewport will result in a different rasterized heightmap. Thus offset features will wobble as the camera is moved. An extreme example is a thin column terrain layer offseting points. Sometimes the sample will hit and other times miss

https://user-images.githubusercontent.com/453755/216322062-2c6966f7-d94d-4c81-abc5-cc5c62683112.mov

Perhaps the heightmap could be rendered into a viewport defined in the coordinate space of the terrain layer?

Reply:
It was a hard decision to make. There is only one height map for all layers. If we index the height information by common space coordinates instead of clipspace coordinates, then for high-pitch scenarios, the resolution closer to the camera will be very bad. I can try construct a viewport from the pitch, bearing and the common bounds, which could be more stable. The shadow module suffers from the similar problem, and honestly it's never completely solved.

Reply:
It would be great if at least when panning we could avoid the glitches. I know it is a hard problem to solve, one approach I've used in the past was to quantize the viewport calculations (e.g. round to the nearest degree/zoom) so that small changes don't cause a heightmap re-render. While this doesn't completely solve the issue it at least means there are fewer jumps. 

I also have the same issue in the `CollideExtension` where fading isn't completely smooth due to sampling precision. I've considered writing the visibilities into a intermediate buffer/texture and averaging over frames to smooth out the visibility value, but it is too much to fit into the current work, but perhaps something to consider as a future improvement.

Reply:
> If we index the height information by common space coordinates instead of clipspace coordinates, then for high-pitch scenarios, the resolution closer to the camera will be very bad.

Superficially, this seems quite similar to the problems that arise for perspective shadow maps. The good news is that there are a wide range of techniques to address those. 

Rendering the shadowmap in three zones  (top thirds, middle third, bottom third), rendering by object distance, logarithmic shadowmaps etc. It is not hard to find good articles.

If my hunch is right, we might even be able to share such code with an improved shadow effect?

Now this implementation seems to have some nice advantages, it should have good performance characteristics for static layers, however for heavily animated layers or very dense scenes where the draw call count becomes very high with this method, it is not impossible that a more "traditional", non-tiled approach could be useful

Reply:
Thank you both for the suggestions. I'm interested in exploring this idea further in a future PR, especially as a reusable strategy in multiple effects. As it is now the height map texture is not bound to any viewport type (common space -> clip space projection is handled by its own shader, not the standard project module) which means we could potentially use regions of this texture differently to provide multi-tier resolution.

This PR is already huge as is. I acknowledge that various functionalities could still be improved, but I prefer that we land a good skeleton first then improve it incrementally.